### PR TITLE
Upgrade to v2.2.0

### PR DESCRIPTION
**Pull request by riversyang**
## All Changes

* Add property `appchain_template_type` to contract struct. The type is defined as:

```rust
pub enum AppchainTemplateType {
    Barnacle,
    BarnacleEvm,
}
```

* Change length checking condition of account id in appchain, based on `appchain_template_type`.
* Add field `offenders` to appchain message `EraPayout`. The types are defined as:

```rust
pub struct Offender {
    pub kind: String,
    pub who: String,
    pub offences: u32,
}

pub struct EraPayoutPayload {
    pub end_era: u32,
    pub excluded_validators: Vec<String>,
    pub offenders: Vec<Offender>,
}
```

* Change data type of `block_height` and `timestamp` in view functions to `U64`.
* Add permissionless function `process_appchain_messages_with_all_proofs`. It is defined as:

```rust
    fn process_appchain_messages_with_all_proofs(
        &mut self,
        signed_commitment: Vec<u8>,
        validator_proofs: Vec<ValidatorMerkleProof>,
        mmr_leaf_for_mmr_root: Vec<u8>,
        mmr_proof_for_mmr_root: Vec<u8>,
        encoded_messages: Vec<u8>,
        header: Vec<u8>,
        mmr_leaf_for_header: Vec<u8>,
        mmr_proof_for_header: Vec<u8>,
    );
```

### CODE REVIEW

File: appchain-anchor/src/appchain_messages.rs [hunk 1]
```diff
@@ -98,8 +98,12 @@ impl IndexedAndClearable for u32 {
         ()
     }
     //
-    fn clear_extra_storage(&mut self) {
-        ()
+    fn clear_extra_storage(&mut self) -> MultiTxsOperationProcessingResult {
```
Comment (hsxyl) [inline appchain-anchor/src/appchain_messages.rs:101]: Why is this function named `clear_extra_storage`?

Comment (riversyang) [reply appchain-anchor/src/appchain_messages.rs:101]: It is for clear extra storage data besides the data of type `T` itself in `LookupArray`. If some collection fields are used in type `T`, the implementation of `T` should implement this function to clear all extra storage keys they used. For example, we may use `LookupMap` or `UnorderedSet` in type `T`, then we should implement this function to clear all storage keys used in these collection types.

Comment (hsxyl) [reply appchain-anchor/src/appchain_messages.rs:101]: Ok, I get it.

File: appchain-anchor/src/appchain_messages.rs [hunk 2]
```diff
@@ -1,13 +1,109 @@
...
+    LockNft,
+}
+
+#[derive(Clone, Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct BurnAssetPayload {
+    token_id: String,
+    sender: String,
+    receiver_id: AccountId,
+    amount: u128,
```
Comment (hsxyl) [inline appchain-anchor/src/appchain_messages.rs:20]: The amount type should be U128.

File: appchain-anchor/src/appchain_messages.rs [hunk 3]
```diff
@@ -1,13 +1,109 @@
...
+}
+
+#[derive(Clone, Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct LockPayload {
+    sender: String,
+    receiver_id: AccountId,
+    amount: u128,
```
Comment (hsxyl) [inline appchain-anchor/src/appchain_messages.rs:28]: The amount type should be U128.

File: appchain-anchor/src/appchain_messages.rs [hunk 4]
```diff
@@ -1,13 +1,109 @@
...
+    pub offenders: Vec<Offender>,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone)]
+#[serde(crate = "near_sdk::serde")]
+pub struct LockNftPayload {
+    pub sender: String,
+    pub receiver_id: AccountId,
+    pub class: u128,
+    pub instance: u128,
```
Comment (hsxyl) [inline appchain-anchor/src/appchain_messages.rs:59]: The amount type should be U128.

File: appchain-anchor/src/appchain_messages.rs [hunk 5]
```diff
@@ -1,13 +1,109 @@
...
+    BurnAsset(BurnAssetPayload),
+    Lock(LockPayload),
+    PlanNewEra(PlanNewEraPayload),
+    EraPayout(EraPayoutPayload),
+    LockNft(LockNftPayload),
+}
+
+#[derive(Encode, Decode, Clone)]
+pub struct RawMessage {
+    nonce: u64,
```
Comment (hsxyl) [inline appchain-anchor/src/appchain_messages.rs:84]: The `nonce` type should be U64

File: appchain-anchor/src/appchain_messages.rs [hunk 6]
```diff
@@ -150,3 +249,195 @@ impl AppchainMessages {
...
+                                receiver_id_in_near: payload.receiver_id,
+                                amount: payload.amount.into(),
+                            },
+                        });
+                    }
+                    Err(err) => appchain_messages.insert_processing_result(
+                        raw_message.nonce as u32,
+                        &AppchainMessageProcessingResult::Error {
+                            nonce: raw_message.nonce as u32,
+                            message: format!("Failed to deserialize raw message paylod: {}", err),
```
Comment (hsxyl) [inline appchain-anchor/src/appchain_messages.rs:300]: The 'paylod' should be 'payload'.

Comment (riversyang) [reply appchain-anchor/src/appchain_messages.rs]: Will fix.

**Review by hsxyl — APPROVED** [review]

**Pull request merged**